src/Val.cc

```diff
@@ -3071,10 +3066,21 @@ ValPtr RecordVal::DoClone(CloneState* state)
 	return rv;
 	}
 
-unsigned int RecordVal::Footprint(bool count_entries) const
+unsigned int RecordVal::Footprint() const
 	{
-	unsigned int fp = 0;
+	// Which records we're in the process of analyzing - used to
+	// avoid infinite recursion for circular types (which can only
+	// occur due to the presence of records).
+	static std::unordered_set<const RecordVal*> pending_records;
```

Comment:
Declaring it `thread_local` instead of `static` could maybe be an alternative that requires little code change and still avoids threading bugs.

Reply:
It's not just threading, also loops that could happen in other ways (I admit that's unlikely here, but I'd rather avoid the pattern of using statics).

Reply:
Okay, I've pushed a fix for this that switches to a stack-based approach.

BTW, I looked into the ASAN failure and it's because the BTest I added includes your mutually-recursive records, and those do indeed induce leaks (seems unavoidable).  So we should either skip ASAN for that btest, or add those leaks to its whitelist.  Not sure how to do either of those.

Reply:
(whoops I missed your comment above about maintaining a simpler interface - my latest push adds that)

Reply:
Yeah, I think we'll need to disable/ignore the test if running with ASAN also not sure if we have an easy way to do that already, @Tim might know?

Reply:
Can `btest` check for the existence of an environment variable to avoid running a test? We could easily define something in the cirrus environment if that's the case.

There's ways to exclude specific blocks of code from the sanitizer, but they're all c++-based and all take effect at compile time.

Reply:
Following the other naming, something like `ZEEK_CI_ASAN_SKIP_TEST`? I think the `ASAN` needs to be there to be very specific about what it's doing.